erizo/src/erizo/thread/Scheduler.cpp

```diff
@@ -0,0 +1,90 @@
+#include "thread/Scheduler.h"
+
+#include <assert.h>
+
+#include <boost/bind.hpp>
+#include <utility>
+
+
+Scheduler::Scheduler(int n_threads_servicing_queue)
+: n_threads_servicing_queue_(0), stop_requested_(false), stop_when_empty_(false) {
+  for (unsigned int index = 0; index < n_threads_servicing_queue; index++) {
+    group_.create_thread(boost::bind(&Scheduler::serviceQueue, this));
+  }
+}
+
+Scheduler::~Scheduler() {
+  stop(false);
+  assert(n_threads_servicing_queue_ == 0);
+}
+
+void Scheduler::serviceQueue() {
+  std::unique_lock<std::mutex> lock(new_task_mutex_);
+  ++n_threads_servicing_queue_;
+  stop_requested_ = false;
+  stop_when_empty_ = false;
+
+  while (!stop_requested_ && !(stop_when_empty_ && task_queue_.empty())) {
+    try {
+      while (!stop_requested_ && !stop_when_empty_ && task_queue_.empty()) {
+        new_task_scheduled_.wait(lock);
+      }
+
+      while (!stop_requested_ && !task_queue_.empty() &&
+      new_task_scheduled_.wait_until(lock, task_queue_.begin()->first) != std::cv_status::timeout) {
+      }
+      if (stop_requested_) {
+        break;
+      }
+
+      if (task_queue_.empty())
+      continue;
+
+      Function f = task_queue_.begin()->second;
+      task_queue_.erase(task_queue_.begin());
+
+      lock.unlock();
+      f();
+      lock.lock();
+    } catch (...) {
+      --n_threads_servicing_queue_;
+      throw;
+    }
+  }
+  --n_threads_servicing_queue_;
+}
+
+void Scheduler::stop(bool drain) {
+  {
+    std::unique_lock<std::mutex> lock(new_task_mutex_);
+    if (drain) {
+      stop_when_empty_ = true;
+    } else {
+      stop_requested_ = true;
+    }
+  }
+  new_task_scheduled_.notify_all();
+  group_.join_all();
+}
+
+void Scheduler::schedule(Scheduler::Function f, std::chrono::system_clock::time_point t) {
+  {
+    std::unique_lock<std::mutex> lock(new_task_mutex_);
+    task_queue_.insert(std::make_pair(t, f));
```

Comment:
Considering including a comment or a reminder that the tasks are ordered by time_point in the multimap, to make it easier to contribute and avoid confusions